design/baremetal-operator/maintenance-mode.md

```diff
@@ -0,0 +1,93 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# Add support for Maintenance Mode flag
+
+## Status
+
+implementable
+
+## Summary
+
+Currently there is no way to prevent management of BaremetalHost resources
+after provisioning is completed.  The underlying Ironic API contains a
+maintenance-mode flag which could be used to disable management of
+BMH resources.
+
+
+## Motivation
+
+In a multi-tier deployment where one cluster deploys another, the "parent"
+cluster will contain BMH resources for initial provisioning,
+but the "child" cluster will later contain BMH resources that reference the
+same physical hosts.
+
+In this scenario it's necessary to prevent management operations from the
+parent cluster such as asserting power state, or BMH actions on the child
+cluster such as the [reboot annotation](reboot-interface.md) may fail due
+to unwanted BMC interactions from the parent cluster.
+
+This may also be useful for temporary maintenance operations
+where you may wish for power management operations and polling
+to be stopped for specific hosts (this is the intended use if the
+underlying [Ironic API](https://docs.openstack.org/api-ref/baremetal/?expanded=set-maintenance-flag-detail#set-maintenance-flag)
+
+### Goals
+
+Add a maintenance flag to the BMH API which uses the underlying Ironic
+interface to disable post-deploy management of a host.
+
+### Non-Goals
+
+
+
+
+## Proposal
+
+
+#### Baremetal Operator
+
+Introduce a new field `maintenance` in the spec of the
+BaremetalHost object. This will be a boolean flag, which
+when set to `true` will put the underlying Ironic host
+into maintenance mode.
+
+```yaml
+apiVersion: metal3.io/v1alpha1
+kind: BareMetalHost
+metadata:
+  name: worker-0
+spec:
+  online: true
+  maintenance: true
+ ...
+```
+
+TODO - we don't expose any reason here, is it enough to have
+a standard reason string e.g "disabled by BaremetalHost maintenance field"?
+
+TODO - what happens if a BMH is deleted with `maintenance: true` - do we
+skip deprovisioning or should that be an independent annotation/field?
```

Comment:
We should look at what Ironic does and see if we want that to be our default behavior.

@dtantsur thoughts?

Reply:
From local testing I'm fairly sure that when you've set the maintenance flag you can just delete the node without any deprovisioning, so aligning with that behavior seems reasonable to me.

Reply:
That matches what I remember.